File: 04.jl
day = "04"
lines = open("inputs/$(day).txt") do f
    readlines(f); #
end

lines = sort(lines)

function parseLog(s)
    (s1, s2) = split(s, ":")
    (min, log) = split(s2, "] ")
    gid = split(log, "#")
    if length(gid) > 1
        gid = gid[2]
        gid = split(gid, " ")[1]
        gid = parse(Int, gid)
        event = nothing
    else
        gid = nothing
        event = log
    end
    return (gid, event, parse(Int, min))
end

function fillGuardIDs(gids)
    for (i, g) in enumerate(gids)
        if g == nothing
            gids[i] = gids[i-1]
        end
    end
    gids
end

logs = [parseLog(s) for s in lines]
gids = getindex.(logs, 1)
gids = fillGuardIDs(gids);
events = getindex.(logs, 2)
mins = getindex.(logs, 3);

ugids = unique(gids)
d = Dict([(u, 0) for u in ugids])
dd = Dict([(u, zeros(60)) for u in ugids])

for (i, (g1, e1, m1)) in enumerate(zip(gids, events, mins))
    if e1 == "falls asleep"
        (g2, e2, m2) = (gids[i+1], events[i+1], mins[i+1])
        if (e2 == "wakes up") && (g2 == g1)
            d[g1] += (m2 - m1)
            dd[g1][(1+m1):(m2)] .+= 1
        else
            d[g1] += (60 - m1)
            dd[g1][(1+m1):end] .+= 1
        end
    end
end

sleepy_gid = [k for (k,v) in d if v == maximum(values(d))][1]
sleepy_minute = [i-1 for (i,x) in enumerate(dd[sleepy_gid])
                 if x == maximum(dd[sleepy_gid])][1]
p1 = sleepy_gid * sleepy_minute
print("Part 1: ", sleepy_gid, " * ", sleepy_minute, " = ", )
println(p1)


most_sleeps = maximum([maximum(dd[u]) for u in ugids])
for u in ugids
    if any(dd[u] .== most_sleeps)
        global p2gid, p2min, p2
        p2gid = u
        p2min = findall(x->x==most_sleeps, dd[u])[1] - 1
        p2 = p2gid * p2min
        break
    end
end
println("Part 2: ", p2gid, " * ", p2min, " = ", p2)

File: 04.py
from util import read
import numpy as np
from collections import defaultdict

def parseInfo(s):
    s = s[1:]
    datetime, info = s.split('] ')
    minute = int(datetime.split(':')[1])
    return minute, info

def parseGuardID(log):
    if '#' in log:
        ret = int(log.split('#')[1].split(' ')[0])
    else:
        ret = None
    return ret

def getGuardIDs(logs):
    gid = np.zeros(logs.shape)
    for j in range(logs.size):
        idno = parseGuardID(logs[j])
        if idno is not None:
            gid[j] = idno
        else:
            gid[j] = gid[j-1]
    gid = gid.astype(int)
    return gid

def getMinAndLogs(x):
    x = np.sort(x)
    info = [parseInfo(s) for s in x]
    minutes, logs = list(zip(*info))
    logs = np.array(logs)
    minutes = np.array(minutes)
    return minutes, logs


def countSleepiness(minutes, logs, gids):
    """
    df : DataFrame
    dttm                 log           gid    minute
    <datetime>           <str>         <int>  <int>
    1518-02-16 00:35:00  falls asleep  3557   35
    1518-02-16 00:38:00	 wakes up      3557   38
    1518-02-16 00:43:00	 falls asleep  3557   43
    1518-02-16 00:45:00  wakes up      3557   45
    1518-02-17 00:23:00  falls asleep  3457   23
    ...
    """
    total_minutes_asleep = defaultdict(int)
    minutes_asleep = {i: np.zeros(60) for i in gids}
    for j in range(len(logs)):
        minute, log, gid = minutes[j], logs[j], gids[j]
        if 'falls asleep' in log:
            mminute, llog, ggid = minutes[j+1], logs[j+1], gids[j+1]
            t0 = minute
            if ('wakes up' in llog) and (ggid == gid):
                t1 = mminute
            else:
                t1 = 60
            minutes_asleep[gid][t0:t1] += 1
            total_minutes_asleep[gid] += (t1 - t0)
    return minutes_asleep, total_minutes_asleep


def argmax(dd):
    vmax = np.max([v for v in dd.values()])
    amax = [k for k, v in dd.items() if v == vmax]
    if len(amax) == 1:
        amax = amax[0]
    return amax

def answer1(minutes, logs, gids):
    minutes_asleep, total_minutes_asleep = countSleepiness(minutes, logs, gids)
    sleepy_gid = argmax(total_minutes_asleep)
    sleepiest_minute = np.argmax(minutes_asleep[sleepy_gid])
    return sleepy_gid * sleepiest_minute

def answer2(minutes, logs, gids):
    minutes_asleep, _ = countSleepiness(minutes, logs, gids)
    most_sleeps = np.max([v.max() for v in minutes_asleep.values()])
    sleepiest_gid = [k for k,v in minutes_asleep.items()
                     if np.any(v == most_sleeps)]
    assert len(sleepiest_gid) == 1, f"more than one gid found: {sleepiest_gid}"
    sleepiest_gid = sleepiest_gid[0]
    sleepiest_minute = np.argmax(minutes_asleep[sleepiest_gid])
    return sleepiest_minute * sleepiest_gid


if __name__ == "__main__":
    day = '04'
    print(f'Day {day}')
    print('------')
    print('Part 1: ', end='')
    x = read(day)
    minutes, logs = getMinAndLogs(x)
    gids = getGuardIDs(logs)

    print(answer1(minutes, logs, gids))
    print('\nPart 2: ', end='')
    print(answer2(minutes, logs, gids))

# Part 1: 138280
